src/common/perf_counters_cache_key.h

```diff
@@ -0,0 +1,118 @@
+                                     std::initializer_list<label_pair> labels);
+
+// TODO: cache_key_insert() to add more labels to an existing string
+
+
+namespace detail {
+
+// use a null character to delimit strings
+constexpr char delimiter = '\0';
+
+// write a string and null delimiter to the output
+constexpr auto insert(std::string_view str,
+                      std::output_iterator<char> auto out)
+{
+  out = std::copy(str.begin(), str.end(), out);
+  *out = delimiter;
+  return std::next(out);
+}
+
+constexpr auto parse_label(std::contiguous_iterator auto pos,
+                           std::contiguous_iterator auto end,
+                           label_pair& label)
+{
+  auto d = std::find(pos, end, delimiter);
+  label.first = std::string_view{pos, d};
+  if (d != end) {
+    auto pos2 = std::next(d);
+    d = std::find(pos2, end, delimiter);
+    label.second = std::string_view{pos2, d};
+  }
+  return d;
+}
+
+constexpr std::size_t label_size(const label_pair& p)
+{
+  return p.first.size() + sizeof(delimiter)
+      + p.second.size() + sizeof(delimiter);
+}
+
+constexpr auto insert_label(const label_pair& label,
+                            std::contiguous_iterator auto begin,
+                            std::contiguous_iterator auto end)
+{
+  // linear search for the upper bound of 'label'
+  struct {
+    std::string::iterator pos;
+    label_pair label;
+  } upper_bound;
+
+  auto pos = begin;
+  do {
+    if (pos == end) {
+      // no upper bound, insert at the end
+      end = insert(label.first, end);
+      end = insert(label.second, end);
+      return end;
+    }
+    upper_bound.pos = pos;
+    pos = parse_label(pos, end, upper_bound.label);
+  } while (upper_bound.label < label);
+
+  // shift remaining buffer contents to the right to make room
+  const auto insert_size = label_size(label);
+  std::advance(end, insert_size);
+  std::shift_right(upper_bound.pos, end, insert_size);
+
+  // copy the label
+  pos = insert(label.first, upper_bound.pos);
+  pos = insert(label.second, pos);
+
+  return end;
+}
+
+} // anonymous namespace
+
+constexpr std::string make_cache_key(std::string_view counter_name,
+                                     std::initializer_list<label_pair> labels)
+{
+  // calculate the total size and preallocate the buffer
+  auto size = std::accumulate(labels.begin(), labels.end(),
+                              counter_name.size() + sizeof(detail::delimiter),
+                              [] (std::size_t sum, const label_pair& label) {
+                                return sum + detail::label_size(label);
+                              });
+  std::string result;
+  result.resize(size);
+
+  // copy the counter_name. 'begin' marks the first label set element
+  const auto begin = detail::insert(counter_name, result.begin());
+  auto end = begin;
+
+  // insert each label into its sorted position
+  for (const auto& label : labels) {
```

Comment:
So, I guess, one question. Would it work to sort the initializer list then just append?

Then we could save the linear-search-and-append for inserting more labels into already existing cache keys.

Reply:
unfortunately initializer_list only provides const iterators. i don't know of any way to sort in-place

Reply:
That sounds like a good approach to me.